Add support ospi driver for Renesas RA8 devices

thaoluonguw opened the pull request:
Add support for ospi driver running on Renesas RA8 (RA8M1 and RA8D1)

- boards: renesas: Add support for ospi on ek_ra8m1 and ek_ra8d1
- drivers: flash: Add support for ospi_b driver
- dts: arm: Add support for ospi
- dts: bindings: Add support for ospi_b
- tests: drivers: flash: common: Add support for ospi.
- samples: drivers: jesd216: Add support for ospi
- samples: drivers: spi_flash: Add support for ospi

Note: This PR requires https://github.com/zephyrproject-rtos/zephyr/pull/79766 is merged

Commit:
drivers: flash: Initial support OSPI flash driver on RA8 boards

Support OSPI flash driver on EK-RA8M1 and EK-RA8D1 with ospi_b
and S28HL512T flash.

Signed-off-by: Tri Nguyen <tri.nguyen.wj@bp.renesas.com>
Signed-off-by: Thao Luong <thao.luong.uw@renesas.com>

GeorgeCGV commented on dts/arm/renesas/ra/ra8/ra8x1.dtsi:
```diff
@@ -473,6 +473,15 @@
 			status = "disabled";
 		};
 
+		ospi0: ospi@40268000 {
```
Consider using:
 
```suggestion
		ospi0: spi@40268000 {
```

to avoid

> node name for SPI buses  should be 'spi'

quytranpzz replied:
@GeorgeCGV : thanks for your comment and sorry for the late response, 
just a confirmation, may I ask about the app you are building with that CMake warning? I didn't encounter that issue on my side.

GeorgeCGV replied:
Based on other targets the general rule is to keep spi buses named spi: https://github.com/zephyrproject-rtos/zephyr/pull/78972, https://github.com/zephyrproject-rtos/zephyr/issues/24499, etc...
However, the sample building for the provided `ek_ra8d1` and `ek_ra8m1` boards doesn't trigger the warning. I am not sure why there is such inconsistency in DTC output.

quytranpzz committed:
dts: renesas: Add devicetree property for OSPI support on RA8

Add ospi node on Renesas RA8 devicetree to support QSPI flash driver

Signed-off-by: Quy Tran <quy.tran.pz@renesas.com>

quytranpzz committed:
boards: renesas: Add boards support for OSPI flash driver on RA8

Add support for OSPI flash driver on EK-RA8D1 and EK-RA8M1

Signed-off-by: Quy Tran <quy.tran.pz@renesas.com>

thaoluonguw commented:
Hello @GeorgeCGV : Could you please help us review again? Thank you so much.

GeorgeCGV commented on dts/bindings/flash_controller/renesas,ra-ospi-b-nor.yaml:
```diff
@@ -0,0 +1,70 @@
+on-bus: ospi
+
+properties:
+  reg:
+    required: true
+    description: Flash Memory base address and size in bytes
+
+  size:
+    type: int
+    required: true
+    description: Flash Memory size
```
The code uses `DT_PROP(RA_OSPI_B_NOR_NODE, size)` but bindings provide reg and size to set the size. Not clear what needs to be used without looking into the driver's code.

Some drivers started to use `reg` to provide the size not a long time ago. In that case, the code has to be modified and the `size` can be removed.

thaoluonguw replied:
Thank you for your feedback.
For `get the size from the reg and remove the size`, I cannot get the size from reg with unknown reason. Building will be failed; it only recognizes to get the address.
I updated to include flash-controller.yaml and change size to size-in-bytes to easy aware the unit.

GeorgeCGV replied:
The idea was to mention the unit within the binding description, i.e., "Flash memory size in bytes," rather than renaming the property.

You can check out #76735 for the `reg` approach.

thaoluonguw replied:
@GeorgeCGV : I updated as you suggested. Thank you.

GeorgeCGV commented on drivers/flash/flash_renesas_ra_ospi_b.h:
```diff
@@ -0,0 +1,358 @@
+#include "spi_nor.h"
+
+/* Device node */
+#define RA_OSPI_B_NOR_NODE DT_INST(0, renesas_ra_ospi_b_nor)
+
+/* Flash size, Write size */
+#define FLASH_SIZE   DT_PROP(RA_OSPI_B_NOR_NODE, size)
+#define WRITE_BLK_SZ DT_PROP(RA_OSPI_B_NOR_NODE, write_block_size)
+
+/* Protocol Mode */
+#define PROTOCOL_MODE DT_PROP(RA_OSPI_B_NOR_NODE, protocol_mode)
+
+/* Data Rate */
+#define DATA_RATE DT_PROP(RA_OSPI_B_NOR_NODE, data_rate)
+
+/* Max frequency */
+#define MAX_FREQUENCY DT_PROP(RA_OSPI_B_NOR_NODE, ospi_max_frequency)
```
Most drivers don't use extra defines for that and use `DT_PROP` directly in the instance.
That makes it easier to read a bit (_no need to go back and forth_).

thaoluonguw replied:
Thank you for your feedback, I updated as you suggest.

GeorgeCGV commented on drivers/flash/flash_renesas_ra_ospi_b.h:
```diff
@@ -0,0 +1,358 @@
+				     .dummy_cycles = DUMMY_CYCLE_WRITE_OSPI},
+};
+
+#define ZEPHYR_ERROR(err)                                                                          \
+	{                                                                                          \
+		if (err != 0) {                                                                    \
+			err = -EIO;                                                                \
+			break;                                                                     \
+		}                                                                                  \
+	}
```
Slightly complicates the code reading and might lead to errors. The name doesn't say anything regarding `break` usage.

Why not use that directly in the driver's code (_where required_)?
Many drivers use `label` and `go to` statements to handle errors and unlock a `mutex` in case of an error.

thaoluonguw replied:
Thank you for your feedback. I updated it. I removed the macro and checked error in the code.
From my viewpoint, I don't prefer using goto.

GeorgeCGV commented on drivers/flash/flash_renesas_ra_ospi_b.c:
```diff
@@ -0,0 +1,754 @@
+	switch (code) {
+	case FLASH_EX_OP_RESET:
+		err = flash_renesas_ra_ospi_b_write_enable(&ospi_b_data->ospi_b_ctrl);
+		ZEPHYR_ERROR(err);
```
Like here, could be a `goto` a label outside of the switch that unlocks the device and returns an error. It is not clear what `ZEPHYR_ERROR` does without looking into it.

thaoluonguw replied:
Thank you for your feedback. I updated it.

GeorgeCGV commented:
> Hello @GeorgeCGV : Could you please help us review again? Thank you so much.

Hello, great work overall. Unfortunately, I don't have the platform to validate fully. I left some comments and concerns.
Also, the initial value assignment (e.g., `int err = 0`) can be omitted in some places.

thaoluonguw commented on drivers/flash/flash_renesas_ra_ospi_b.c:
```diff
@@ -0,0 +1,754 @@
+	acquire_device(dev);
+
+	switch (code) {
+	case FLASH_EX_OP_RESET:
```
Thank you for your feedback, I updated it.

GeorgeCGV replied:
The proposal was more about avoiding unnecessary lock & unlock:

```c
int err;

ARG_UNUSED(in);
ARG_UNUSED(out);

if (code != FLASH_EX_OP_RESET) {
	return -ENOTSUP;
}

acquire_device(dev);
      ...
release_device(dev);

return err;
```

GeorgeCGV commented on drivers/flash/flash_renesas_ra_ospi_b.h:
```diff
@@ -0,0 +1,337 @@
+/* Calibration sector */
+#define SECTOR_THREE (2U)
+
+/* Flash opcodes */
+#define SPI_NOR_CMD_WR_REG2  (0x71) /* Write config register 2 */
+#define SPI_NOR_CMD_RSFDPID  (0x5A) /* Read SFDP ID */
+#define SPI_NOR_CMD_RREG     (0x65) /* Read Any Register */
+#define SPI_NOR_CMD_SE_4KB   (0x21) /* Sector Erase 4Kb address */
+#define SPI_NOR_CMD_SE_256KB (0xDC) /* Sector Erase 128Kb and 256Kb address */
+#define SPI_NOR_CMD_DE       (0x60) /* Chip Erase */
+
+/* Flash octal opcodes */
+#define SPI_NOR_OCMD_WEN      (0x0606) /* Octal Write enable */
+#define SPI_NOR_OCMD_RSR      (0x0505) /* Octal Read status register */
+#define SPI_NOR_OCMD_WR_REG2  (0x7171) /* Octal Write config register 2 */
+#define SPI_NOR_OCMD_RDID     (0x9F9F) /* Octal Read JEDEC ID */
+#define SPI_NOR_OCMD_RSFDPID  (0x5A5A) /* Octal Read SFDP ID */
+#define SPI_NOR_OCMD_RREG     (0x6565) /* Octal Read Any Register */
+#define SPI_NOR_OCMD_PP_4B    (0x1212) /* Octal Page Program 4 Byte Address */
+#define SPI_NOR_OCMD_READ     (0xEEEE) /* Octal Read data */
+#define SPI_NOR_OCMD_RST_EN   (0x6666) /* Octal Reset Enable */
+#define SPI_NOR_OCMD_RST_MEM  (0x9999) /* Reset Memory */
+#define SPI_NOR_OCMD_SE_4KB   (0x2121) /* Octal Sector Erase 4Kb address */
+#define SPI_NOR_OCMD_SE_256KB (0xDCDC) /* Octal Sector Erase 128Kb and 256Kb address */
+#define SPI_NOR_OCMD_DE       (0x6060) /* Octal Chip Erase */
```
Some `SPI_NOR_` definitions exist in [drivers/flash/spi_nor.h](https://github.com/zephyrproject-rtos/zephyr/blob/f3bf6357c799191c230142ecb04994c0f232ed16/drivers/flash/spi_nor.h) and missing ones can be added there.

thaoluonguw replied:
I updated as you suggested, could you please re-check? Thank you.

GeorgeCGV reviewed (dismissed):
@thaoluonguw, thank you for the adaptations. It would be great to factor out the `spi_nor.h` modification and reformat to a standalone commit.

By the way, are you aware of the #71769 effort? It might have been beneficial to implement the new driver against that API.

khoa-nguyen-18 commented:
Just rebased the main branch to resolve conflicts. The PR will need some updates, which will be pushed soon.

khoa-nguyen-18 committed:
samples: drivers: jesd216: Add support OSPI for Renesas RA8

Add support OSPI for Renesas ek_ra8m1, ek_ra8d1 to run
sample jesd216

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>
Signed-off-by: Tri Nguyen <tri.nguyen.wj@bp.renesas.com>
Signed-off-by: Thao Luong <thao.luong.uw@renesas.com>

khoa-nguyen-18 committed:
samples: drivers: spi_flash: Add support OSPI for Renesas RA8

Add support OSPI for Renesas ek_ra8m1, ek_ra8d1 to run sample
spi_flash

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>
Signed-off-by: Tri Nguyen <tri.nguyen.wj@bp.renesas.com>
Signed-off-by: Thao Luong <thao.luong.uw@renesas.com>

khoa-nguyen-18 commented:
- Seperate the `spi_nor.h` modification and reformating to another commit
- Correct the naming of `Kconfig.renesas_ra_flash_ospi_b` file and `FLASH_OSPI_B_RENESAS_RA` config
- Modify commit support test `flash/common` to add test case for OSPI

khoa-nguyen-18 commented:
> @thaoluonguw, thank you for the adaptations. It would be great to factor out the `spi_nor.h` modification and reformat to a standalone commit.

Hi @GeorgeCGV, sorry to bother you again after a long time. I just rebased to resolve conflicts and added a minor modification that I mentioned above. Could you please take a look at this PR?

khoa-nguyen-18 commented:
rebase to solve conflicts

GeorgeCGV commented:
@khoa-nguyen-18 please avoid extensive reformatting of `spi_nor.h` to ensure commit [d150f56f2b3f7da718c317a11538a820fd8ef7ec](https://github.com/zephyrproject-rtos/zephyr/pull/80799/commits/d150f56f2b3f7da718c317a11538a820fd8ef7ec) reflects actual modifications and new lines.

khoa-nguyen-18 commented:
rebase main to solve conflicts

binhnguyen2434 commented on boards/renesas/ek_ra8d1/ek_ra8d1.dts:
```diff
@@ -113,6 +114,17 @@
 		status = "okay";
 	};
 };
+&pll2 {
```
There should be a new line above

khoa-nguyen-18 replied:
Thank you, I have updated this line

binhnguyen2434 commented on boards/renesas/ek_ra8d1/ek_ra8d1.dts:
```diff
@@ -342,3 +360,43 @@ pmod_sd_shield: &sdhc1 {};
+	pinctrl-0 = <&ospi0_default>;
+	pinctrl-names = "default";
+	status = "okay";
+	s28hl512t: ospi-nor-flash@90000000 {
```
Similar

khoa-nguyen-18 replied:
I have updated it

binhnguyen2434 commented on boards/renesas/ek_ra8d1/ek_ra8d1.dts:
```diff
@@ -342,3 +360,43 @@ pmod_sd_shield: &sdhc1 {};
+		ospi-max-frequency = <DT_FREQ_M(200)>;
+		reg = <0x90000000 DT_SIZE_M(64)>;
+		write-block-size = <1>;
+		pages_layout: pages_layout {
+			pages_layout_4k: pages_layout_4k {
+				pages-count = <32>;
+				pages-size = <DT_SIZE_K(4)>;
+			};
+			pages_layout_128k: pages_layout_128k {
+				pages-count = <1>;
+				pages-size = <DT_SIZE_K(128)>;
+			};
+			pages_layout_256k: pages_layout_256k {
+				pages-count = <255>;
+				pages-size = <DT_SIZE_K(256)>;
+			};
```
Similar

khoa-nguyen-18 replied:
I have updated it

khoa-nguyen-18 commented on boards/renesas/ek_ra8m1/ek_ra8m1-pinctrl.dtsi:
```diff
@@ -212,4 +212,22 @@
+			<RA_PSEL(RA_PSEL_OSPI, 8, 2)>, <RA_PSEL(RA_PSEL_OSPI, 8, 4)>;
+			drive-strength = "highspeed-high";
+		};
+		group2 {
```
I have updated it one also

khoa-nguyen-18 commented on boards/renesas/ek_ra8m1/ek_ra8m1.dts:
I have updated it, thank you

binhnguyen2434 reviewed (dismissed):
LGTM!

khoa-nguyen-18 commented on drivers/flash/flash_renesas_ra_ospi_b.c:
```diff
@@ -0,0 +1,775 @@
+		return 0;
+	}
+
+	if (data != NULL) {
+		p_src = data;
+	} else {
+		return -EINVAL;
+	}
```
Hi @de-nordic, many thanks for your reviews, we really appreciate it. Our team will discuss your suggestions and ensure they align with the other driver. Would it be okay if we update your comments and suggestions in a separate PR?

khoa-nguyen-18 replied:
We use `p_src` to cast from `const void *` to `const uint8_t *` as input for the HAL API

khoa-nguyen-18 committed:
tests: drivers: flash: Add support "common" for Renesas OSPI_B

Add support test app "flash/common" for testing Renesas RA
OSPI_B on ek_ra8m1, ek_ra8d1

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 commented on drivers/flash/flash_renesas_ra_ospi_b.c:
```diff
@@ -0,0 +1,775 @@
+	if (!flash_renesas_ra_ospi_b_is_valid_address(dev, offset, len)) {
+		LOG_ERR("address or size exceeds expected values: "
+			"addr 0x%lx, size %zu",
+			(long)offset, len);
```
Thank you. After consideration, we use `LOG_DBG` in the child functions and `LOG_ERR` in the driver API.

khoa-nguyen-18 commented on drivers/flash/flash_renesas_ra_ospi_b.c:
```diff
@@ -0,0 +1,775 @@
+	err = flash_renesas_ra_ospi_b_write_enable(p_ctrl);
+	if (err != FSP_SUCCESS) {
+		return err;
+	}
```
I have updated the code to add `LOG_DBG` for all `return err` statements.

soburi commented on tests/drivers/flash/common/src/main.c:
```diff
@@ -12,6 +12,8 @@
 #if defined(CONFIG_NORDIC_QSPI_NOR)
 #define TEST_AREA_DEV_NODE	DT_INST(0, nordic_qspi_nor)
+#elif defined(CONFIG_FLASH_RENESAS_RA_OSPI_B)
+#define TEST_AREA_DEV_NODE DT_INST(0, renesas_ra_ospi_b_nor)
```
It would be better to align the indentation.

khoa-nguyen-18 replied:
thank you, I have updated it

soburi commented on tests/drivers/flash/common/src/main.c:
```diff
@@ -32,12 +34,18 @@
 
 #define TEST_AREA_DEVICE	DEVICE_DT_GET(TEST_AREA_DEV_NODE)
+#if defined CONFIG_FLASH_RENESAS_RA_OSPI_B
+#define TEST_AREA_OFFSET 0x40000
```
It would be better to align the indentation.

khoa-nguyen-18 replied:
thank you, I have updated it also

khoa-nguyen-18 commented:
rebase main to solve conflict

KhiemNguyenT reviewed (dismissed):
LGTM

de-nordic commented on drivers/flash/spi_nor.h:
```diff
@@ -61,6 +61,12 @@
 #define SPI_NOR_CMD_PP_1_1_4_4B  0x34  /* Quad Page program (1-1-4) 4 Byte Address */
 #define SPI_NOR_CMD_PP_1_4_4_4B  0x3e  /* Quad Page program (1-4-4) 4 Byte Address */
 
+#define SPI_NOR_CMD_WR_WRARG 0x71 /* Write Any Register for S28Hx512T */
+#define SPI_NOR_CMD_RSFDPID  0x5A /* Read SFDP ID for S28Hx512T */
+#define SPI_NOR_CMD_RREG     0x65 /* Read Any Register for S28Hx512T */
+#define SPI_NOR_CMD_SE_256KB 0xDC /* Sector Erase 256KB for S28Hx512T */
+#define SPI_NOR_CMD_ERCHP    0x60 /* Erase Chip for S28Hx512T */
+
```
No. it fit his is S28Hx512T specific, it does not land here. Get your own header for that. Specifically that nothing of this is used in spi_nor.c

de-nordic commented on drivers/flash/spi_nor.h:
```diff
@@ -76,6 +82,20 @@
 #define SPI_NOR_OCMD_RD_CFGREG2 0x718E  /* Octal Read configuration Register2 */
 #define SPI_NOR_OCMD_BULKE      0x609F  /* Octa Bulk Erase */
 
+#define SPI_NOR_OCMD_WEN      0x0606 /* Octal Write enable for S28Hx512T */
+#define SPI_NOR_OCMD_RSR      0x0505 /* Octal Read status register for S28Hx512T */
+#define SPI_NOR_OCMD_WR_REG2  0x7171 /* Octal Write config register 2 for S28Hx512T */
+#define SPI_NOR_OCMD_RDID     0x9F9F /* Octal Read JEDEC ID for S28Hx512T */
+#define SPI_NOR_OCMD_RSFDPID  0x5A5A /* Octal Read SFDP ID for S28Hx512T */
+#define SPI_NOR_OCMD_RREG     0x6565 /* Octal Read Any Register for S28Hx512T */
+#define SPI_NOR_OCMD_PP_4B    0x1212 /* Octal Page Program 4 Byte Address for S28Hx512T */
+#define SPI_NOR_OCMD_READ     0xEEEE /* Octal Read data for S28Hx512T */
+#define SPI_NOR_OCMD_RST_EN   0x6666 /* Octal Reset Enable for S28Hx512T */
+#define SPI_NOR_OCMD_RST_MEM  0x9999 /* Reset Memory for S28Hx512T */
+#define SPI_NOR_OCMD_SE_4KB   0x2121 /* Octal Sector Erase 4Kb address for S28Hx512T */
+#define SPI_NOR_OCMD_SE_256KB 0xDCDC /* Octal Sector Erase 256Kb address for S28Hx512T */
+#define SPI_NOR_OCMD_ERCHP    0x6060 /* Octal Erase Chip for S28Hx512T */
```
Same here.

khoa-nguyen-18 replied:
fixed, thank you

khoa-nguyen-18 commented:
hi @de-nordic, could you please confirm your requested change?

de-nordic commented:
> hi @de-nordic, could you please confirm your requested change?

I re-assigning the PR to @KhiemNguyenT as Renesas representative.

KhiemNguyenT reviewed (approved):
Thank you. LGTM.

bricle commented:
@thaoluonguw @duynguyenxa @KhiemNguyenT  @thenguyenyf  Hello, dear RENESAS dev team! May i ask is there a plan to extend OSPI driver support to new RA8 devices such as RA8P1 and RA8D2, currently these new parts lack on-board ospi nor flash driver support. Thanks a lot for any feedback!

thenguyenyf commented:
> @thaoluonguw @duynguyenxa @KhiemNguyenT @thenguyenyf Hello, dear RENESAS dev team! May i ask is there a plan to extend OSPI driver support to new RA8 devices such as RA8P1 and RA8D2, currently these new parts lack on-board ospi nor flash driver support. Thanks a lot for any feedback!

@bricle It was sent out recently. Please visit https://github.com/zephyrproject-rtos/zephyr/pull/108730

bricle commented:
> > @thaoluonguw @duynguyenxa @KhiemNguyenT @thenguyenyf Hello, dear RENESAS dev team! May i ask is there a plan to extend OSPI driver support to new RA8 devices such as RA8P1 and RA8D2, currently these new parts lack on-board ospi nor flash driver support. Thanks a lot for any feedback!
> 
> @bricle It was sent out recently. Please visit #108730

Thanks! This is very fast.👍
